insights/core/spec_factory.py

```diff
@@ -736,6 +736,61 @@ def __call__(self, broker):
                 keep_rc=self.keep_rc, ds=self, timeout=self.timeout, inherit_env=self.inherit_env)
 
 
+class command_with_args(object):
+    """
+    Execute a command that has dynamic arguments
+
+    Args:
+        cmd (list of lists): the command to execute. Breaking apart a command
+            string that might require arguments.
+        provider (str or list): argument string or a list of arguments.
+        context (ExecutionContext): the context under which the datasource
+            should run.
+        split (bool): whether the output of the command should be split into a
+            list of lines
+        keep_rc (bool): whether to return the error code returned by the
+            process executing the command. If False, any return code other than
+            zero with raise a CalledProcessError. If True, the return code and
+            output are always returned.
+        timeout (int): Number of seconds to wait for the command to complete.
+            If the timeout is reached before the command returns, a
+            CalledProcessError is raised. If None, timeout is infinite.
+        inherit_env (list): The list of environment variables to inherit from the
+            calling process when the command is invoked.
+
+    Returns:
+        function: A datasource that returns the output of a command that takes
+            specified arguments passed by the provider.
+    """
+
+    def __init__(self, cmd, provider, context=HostContext, deps=[], split=True, keep_rc=False, timeout=None, inherit_env=[], **kwargs):
```

Comment:
@xiangce I had a discussion about this with @csams the other day.  For args where the default is a mutable object like `[]` or `{}` there will only be one instance of it for all objects so its better to use the following pattern:
```
def __init__(self, some_arg=None):
    some_arg = some_arg if some_arg is not None else []
```

See [this article](https://blog.florimond.dev/python-mutable-defaults-are-the-source-of-all-evil) for additional info.

Reply:
Ack, however, do we need to modify them in other helpers?